Add  a developer guide document

dcbaker opened the pull request:
This is still largely a work in progress, but provided as a starting point for further discussion

dcbaker commented:
@bretbrownjr, we've had few style/format/etc discussions. I thought it might be easier to open a PR with some initial thoughts and hammer that out than discuss them in feature PRs

bretbrownjr commented on docs/dev.md:
```diff
@@ -0,0 +1,143 @@
+## Editors
+
+Use [editorconfig](https://editorconfig.org). Many editors support editorconfig
+natively, others have plugins. Otherwise, configure your editor to use the same
```
```suggestion
natively. Others have plugins. Otherwise, configure your editor to use the same
```

bretbrownjr commented on docs/dev.md:
```diff
@@ -0,0 +1,143 @@
+
+- external_dependencies (found by `dependency()` or `compiler.find_library()`)
+  should be prefixed with `dep_`, ex: `dep_fmt`
+- internal dependencies (`declare_dependency()`) should be prefixed with `idep_`
```
Just to catch me up... why? In my mind it's cleaner for each unit of code, even subdirectories, to just treat dependencies like dependencies.

dcbaker replied:
I prefer to use the `idep_` identifier in part for historical reasons. But probably doesn't make sense anymore, so I'll change that.

If you're into history, read on:

In ye old days, a consumer of a subproject would make as series of calls like this, pull a named variable from a subproject into their project:
```meson
sub_gtest = subproject('gtest')
dep_gtest = sub_gtest.get_variable('dep_gtest')
```

In the modern approach, a producer does something like:
```meson
dep_gtest = declare_dependency(...)
meosn.override_dependency('gtest', dep_gtest)
```
and then the consumer just does:
```meson
dep_gtest = dependency('gtest')
```
and Meson will do the right thing.

bretbrownjr commented on docs/dev.md:
```diff
@@ -0,0 +1,143 @@
+- internal dependencies (`declare_dependency()`) should be prefixed with `idep_`
+- external programs (`find_program()`) should be prefixed with `prog_`
+
+use spaces around `:`, ex: `dependency('gtest', version : '>= 1.5')`
```
```suggestion
Use spaces around `:`. For example: `dependency('gtest', version : '>= 1.5')`
```

bretbrownjr commented on docs/dev.md:
```diff
@@ -0,0 +1,143 @@
+
+use spaces around `:`, ex: `dependency('gtest', version : '>= 1.5')`
+
+always drop arguments to a new line when they do not fit on one:
```
```suggestion
Always put arguments on a new line when they do not fit on one line:
```

bretbrownjr commented on docs/dev.md:
````diff
@@ -0,0 +1,143 @@
+)
+```
+
+## C and C++
````
I was thinking we'd just pick an existing coding standard? Then we could just say "we use the LLVM coding standard", set up an analyzer, and move on?

bretbrownjr commented on docs/dev.md:
```diff
@@ -0,0 +1,143 @@
+
+Prefer C++ (`//`) comments to C-style (`/* */`) in general.
+
+C style comments may be used in the following situations:
+  - spec citations
+  - algorithm descriptions
```
I understand what this text says. I really don't understand the reasoning behind it though.

bretbrownjr commented on docs/dev.md:
````diff
@@ -0,0 +1,143 @@
+void do_thing(bool x, bool y, bool z);
+```
+
+`#defines` should be all uppercase, and be prefaced with `CPS_CONFIG_`. such as
+`CPS_CONFIG_USE_UNDEFINED`
````
```suggestion
`#defines` should be all uppercase, and be prefaced with `CPS_CONFIG_`. such as
`CPS_CONFIG_USE_UNDEFINED`
```

Well, this is subtle. If we're going to ship a `cps` library, the prefix would be `CPS_`. The executable could also use `CPS_*` because it's shipping in the same repo so collisions are unlikely-to-impossible. Or it could use `CPS_CONFIG_` if we want to treat `cps` and `cps-config` as two projects sharing one build system and repo.

I'm flexible, but I did want to point out that `CPS_CONFIG_FOOBAR` would be odd in a `cps/foobar.hpp`.

bretbrownjr commented on docs/dev.md:
```diff
@@ -0,0 +1,143 @@
+
+## Tests
+
+Complex implementation details may use compiled unit tests, like gtests. Which
+are generally faster and easier to debug than functional tests.
```
```suggestion
Complex implementation details may use compiled unit tests, like gtests. Unit tests
are generally faster and easier to debug than functional tests.
```

bretbrownjr commented on docs/dev.md:
```diff
@@ -0,0 +1,143 @@
+Complex implementation details may use compiled unit tests, like gtests. Which
+are generally faster and easier to debug than functional tests.
+
+Functional tests should use the tests/runner.py, by adding test cases to the
+`tests/cases.toml` along with adding or modifying the provided `.cps` files.
```
```suggestion
Functional tests should use `tests/runner.py` by adding test cases to
`tests/cases.toml`  along with changes to the accompanying `.cps` files.
```

bretbrownjr commented:
I'd be interested if @lunacd or @tylerjw had thoughts as well, given they're jumping in fresh and recently.

lunacd commented on docs/dev.md:
```diff
@@ -0,0 +1,143 @@
+
+## All files
+
+All files should have an [SPDX license
```
It would be helpful to qualify what files would need a SPDX license header and what others wouldn't, unless the intention here is to say literally all files. In that case we should just fix the existing ones that don't have this header, e.g. `meson.build`, all the cps files used for unit tests, etc.

lunacd commented on docs/dev.md:
```diff
@@ -0,0 +1,143 @@
+Complex implementation details may use compiled unit tests, like gtests. Which
+are generally faster and easier to debug than functional tests.
+
+Functional tests should use the tests/runner.py, by adding test cases to the
```
Off topic, and I'm happy to open a separate issue if this calls for its own discussion. Apart from unit and function tests, would it add value to this project to add some more integrative tests, e.g. can we build an interesting project with `cps-config` and some hand-crafted cps files

tylerjw commented on docs/dev.md:
```diff
@@ -0,0 +1,143 @@
+
+## Editors
+
+Use [editorconfig](https://editorconfig.org). Many editors support editorconfig
```
I'm using sublime text and this setting from your file:

```
insert_final_newline = false
```

is causing sublime text to strip final newlines from files. You can see the result of that in the PR I submitted.

tylerjw commented:
You should add a note here on the version of clang-format you are using. I tried locally setting up the automation I like using with the python tool [pre-commit](https://pre-commit.com/), however it reformatted all the source files. This is because versions of clang-format are not compatible with each-other and I suspect you have a different version installed than I do.

We should also add a formatting ci job that verifies the formatting stays as you'd expect in each PR. I can easily contribute this if I know what version of clang-format you are using.

dcbaker commented:
I'm using -16 ATM, but I could easily go backwards, or go up to -17 with my distro currently

tylerjw commented:
Due to this: https://apt.llvm.org/ it is easy enough for me to get whatever versions I need for each project.

My point was more that it needs to be documented to aid people in contributing to the project.

dcbaker commented:
I pulled in all of the suggested changes and squashed those, then rebased this on #49. I can take this off 49 if that's preferable

tylerjw reviewed (approved):
This looks good to me, it does look like you should rebase this to drop the changes you already merged in the other PR.

dcbaker commented:
Latest version has the following changes:
 - Merged more of Bret's cleanups (thanks!)
 - Changed the comments section to always prefer `//` comments
 - Updated the style guide to say "LLVM style with the following exceptions"
 - Attempted to clarify that source files need copyright information, but data files do not
 - removed the Meson `idep_` recommendation, and suggest only `dep_`
 - Added information about clang-format usage and version
 - changed `CPS_CONFIG_` -> `CPS_`

bretbrownjr reviewed (approved):
It's definitely helpful. Let's land it and evolve it from here.
